Sources/NIO/Bootstrap.swift

```diff
@@ -374,6 +374,31 @@ public final class ClientBootstrap {
         }
     }
 
+    /// Use the existing connected socket file descriptor.
+    ///
+    /// - parameters:
+    ///     - descriptor: The _Unix file descriptor_ representing the connected socket.
+    /// - returns: an `EventLoopFuture<Channel>` to deliver the `Channel` immediately.
+    public func withConnectedSocket(descriptor: Int32) -> EventLoopFuture<Channel> {
```

Comment:
Can we add a method like this to all 3 bootstraps? There's utility in passing already-bound sockets to both `ServerBootstrap` and `DatagramBootstrap`.

Reply:
Added a method to the ServerBootstrap, but not to the datagram. The Datagram bootstrap is missing the connect method, and I'd know to see if that has to be added as well.

Reply:
Sorry, I wasn't clear enough.

I said "a method like this", by which I mean "insert a socket that is already active". In the case of the `ServerBootstrap`, "already active" means "has had `listen` called on it". In the case of the `DatagramBootstrap`, currently "already active" means "has had `bind` called on it". We will eventually also support connected datagram sockets, but for now we do not.

Reply:
@Lukasa, not sure that listening it outside is the best way.

 * if we `listen()` outside of the bootstrap I'd have to extend `BaseSocket` with the `listen()` method, with all the consequences, are you OK with that?
 * listening outside robs the framework off the ability to use its own defined `backlog` parameter, which is not very uniform.
 * the serverSocket already does `listen()` for the primitive socket creation.
 * double listen does not hurt (we can put `try?` around it) if the client happened to initialize it
 * Ultimately, this is all about expectations. If we say that SwiftNIO doesn't strictly require you to listen on it, just bind, the expectation is clear and the use can override it with their own listen on the socket if they need a different backlog parameter that much.

Reply:
Why do you need to extend `BaseSocket`? It seems like you can use the `ServerSocket`, given that the `ServerBootstrap` creates an instance of the concrete subclass.

The problem is that right now `withBoundSocket` does not behave the same was as `bind`. In particular, it fires `channelActive` without having had `listen` called on the socket. As a more general note, our pipeline does not have a `listen` method, so if you pass a non-listening file descriptor to `withBoundSocket` as currently implemented then it will *never* start listening (you have no way to signal that request to it). In NIO land, server sockets follow up their binding with calls to `listen` immediately. If we're having a shortcut around needing to bind a socket for ServerSocketChannel then we need to listen it.

That said, I'm open to having `withBoundSocket` call `listen` at that stage. I agree that it's unlikely that a user will have a socket that is already listening, it's much more likely they have one that's already bound than one that's already listening. So I think if we can just update this so that `withBoundSocket` calls `listen` and write the appropriate test that this works.

Reply:
So the test is there and withBoundSocket is there as well. What else do I need to add to this patch, except for adding the same stuff to the datagram bootstrap?

Reply:
Just that, please! Thanks for being willing to work through this, especially with the time zones slowing stuff down.

Reply:
@Lukasa, Pushed, check it out please!